astronomy/standard_product_3.cpp

```diff
@@ -0,0 +1,310 @@
+﻿#include "astronomy/standard_product_3.hpp"
+
+#include <fstream>
+#include <optional>
+
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_cat.h"
+#include "astronomy/time_scales.hpp"
+#include "base/map_util.hpp"
+#include "glog/logging.h"
+
+namespace principia {
+namespace astronomy {
+namespace internal_standard_product_3 {
+
+using base::FindOrDie;
+using geometry::Displacement;
+using quantities::si::Deci;
+using quantities::si::Kilo;
+using quantities::si::Metre;
+using quantities::si::Second;
+
+StandardProduct3::StandardProduct3(
+    std::filesystem::path const& filename,
+    StandardProduct3::Dialect const dialect) {
+  std::ifstream file(filename);
+  CHECK(file.good()) << filename;
+  std::optional<std::string> line;
+  std::string location;
+  int line_number = 0;
+  auto const read_line = [&file, &filename, &line, &line_number, &location]() {
+    if (!line.has_value()) {
+      line.emplace();
+    }
+    std::getline(file, *line);
+    if (file.fail()) {
+      CHECK(file.eof()) << "non-EOF failure after " << location;
+      line.reset();
+      location = absl::StrCat(filename.string(), " at end of file");
+    } else {
+      ++line_number;
+      location =
+          absl::StrCat(filename.string(), " line ", line_number, ": ", *line);
+    }
+  };
+
+  // The specification uses 1-based column indices, and column ranges with
+  // bounds included.
+  auto const column = [&line, &location, line_number](int index) -> char {
+    CHECK(line.has_value()) << location;
+    CHECK_LT(index - 1, line->size()) << location;
+    return (*line)[index - 1];
+  };
+  auto const columns = [&line, &location](
+                            int first, int last) -> std::string_view {
```

Comment:
`int const` (2x).  Also, let's omit the return type when it's not necessary.

Do we believe that Mac has `string_view`?

Reply:
Done.

> Do we believe that Mac has string_view?

 Sort of...